### ISSUE TITLE
compile error when compiling in home directory

### ISSUE THREAD

**Issue by lego** [crystal-lang/crystal#5659]
This is a bit tricky to provide reproducible steps, but this has happened in a school environment where Crystal was set up and installed. As far as I know, it was set up using tar.gz. It happens consistently.

The environment seems to have only LLVM 3.8 although Crystal seems to report 4.0.
```
$ crystal -v
Crystal 0.24.1 (2017-12-22)

LLVM: 4.0.0
Default target: x86_64-unknown-linux-gnu
```

If my current directory is my home directory, I seem to be unable to compile anything with Crystal. I tried using both a file in the home directory (`hello.cr`) and with `crystal eval`. Both inputs consisted of only `puts "Hello"`. The output was:
```
```

# Submitting bugs

Make sure to review these points before submitting issues - thank you!

- Make sure a similar issue doesn't exist yet: use the search box
- **Include reproducible code**: we should be able to paste it into an editor, compile and run it and get the same error as you. Otherwise it's impossible for us to reproduce the bug.
- Don't **only** use `play.crystal-lang.org` or `carc.in`: code might be lost and the issue will remain incomplete. Write code in the issue itself.
- Reduce code, if possible, to the minimum size that reproduces the bug.
- If all of the above is impossible due to a large project, create a branch that reproduces the bug and point us to it.
- Include Crystal compiler version (`crystal -v`) and OS. If possible, try to see if the bug still reproduces on master.

# Submitting questions

- If the question is something like "how do I do X?", consider using [StackOverflow](http://stackoverflow.com/questions/tagged/crystal-lang)
- If the question is about a shard, consider asking the question in that shard

**Comment by lego** [comment]
(sorry this pre-maturely submitted, still editing)

**Comment by lego** [comment]
(finished editing, sorry for the email spam if you see this! deleting both this comment and the previous one regarding the edit)

**Comment by straight-shoota** [comment]
This could be difficult to reproduce without any further details about the environment.
What exactly is your home path (`echo $HOME`)?

Out of curiosity, which school is using Crystal and for what? =)

**Comment by lego** [comment]
(oops, I thought I put it in the original comment).
```bash
$ echo $HOME
/u3/j5pereira
```

The school isn't quite using it, but it is the University of Waterloo. This is for a compilers course where we can use any language we want, so they made it available at my request. =)

**Comment by RX14** [comment]
@lego can you run `crystal env` please?

**Comment by drosehn** [comment]
It might also be interesting to see the output of `echo $PATH`

**Comment by lego** [comment]
```bash
$ crystal env
CRYSTAL_CACHE_DIR="/u3/j5pereira/.cache/crystal"
CRYSTAL_PATH="/usr/bin/../share/crystal/src:lib"
CRYSTAL_VERSION="0.24.1"

$ echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
```

The cache folder stood out so I poked around it. It appears that compilation fails while the current directory is the home directory if there is no cache populated for the targeted crystal build file (regardless of where the target crystal file is). By removing the cache I'm able to reproduce things this problem.

Something else that stood out is that `/usr/bin/../share/crystal` only contains the folder `src`. Should it also have `lib`?

A few other diagnostics:

The permissions on `$HOME/.cache` and `$HOME/.cache/crystal`
```bash
drwx--S---   41 j5pereir none         4096 Jan 30 13:59 .cache
drwxr-sr-x  3 j5pereir none       4096 Jan 30 13:59 crystal
```

The contents of `$HOME/.cache/crystal/u3-j5pereir-hello.cr` after a failed compilation.
```bash
-rw-r--r-- 1 j5pereir none  3424 Jan 30 14:17 A-rray40I-nt3241.bc
-rw-r--r-- 1 j5pereir none     9 Jan 30 14:17 bc_flags
-rw-r--r-- 1 j5pereir none  5020 Jan 30 14:17 D-ebug5858D-W-A-R-F-5858L-ineN-umbers5858R-egister.bc
-rw-r--r-- 1 j5pereir none 11116 Jan 30 14:17 H-ash40S-ignal4432P-roc40S-ignal4432N-il4141.bc
-rw-r--r-- 1 j5pereir none  2156 Jan 30 14:17 I-O-5858E-rror.bc
-rw-r--r-- 1 j5pereir none 56280 Jan 30 14:17 _main.bc
-rw-r--r-- 1 j5pereir none  2236 Jan 30 14:17 M-ath.bc
-rw-r--r-- 1 j5pereir none  3116 Jan 30 14:17 P-ointer40P-roc40I-nt324432N-il4141.bc
-rw-r--r-- 1 j5pereir none  2508 Jan 30 14:17 S-taticA-rray40U-I-nt84432241.bc
-rw-r--r-- 1 j5pereir none 45212 Jan 30 14:17 U-nicode.bc
```

**Comment by lego** [comment]
I also ran `strace crystal build $HOME/hello.cr` immediately after deleting `.cache/crystal`. I haven't yet groked the results. Hopefully, this may give some indication as to what the binary is looking for and why it's failing.  [crystal-build-strace.log](https://github.com/crystal-lang/crystal/files/1679105/crystal-build-strace.log)

**Comment by jhass** [comment]
@Igo since it's been over two years, do you mind trying again? :)

**Comment by tuxor1337** [comment]
I'm able to reproduce a similar error output, but I'm not sure if it is related. But it might well be, because it's also on a shared host. The background is that I want to compile the crystal project https://github.com/omarroth/invidious as a normal user on a shared host (via ssh). When running the crystal-build command with `--threads 1`, the job is killed at some point because the shared host has restrictions on resource use by normal users. However, I can simply rerun the build command several times until, at some point, it will succeed.

To reproduce the error output in this issue, I have to omit the argument `--threads 1`. I suspect that the host is killing compiler threads and the crystal builder misinterprets this as finished threads. It invokes the linking steps before the compiler is actually finished leading to missing object files.

Details: I install all the dependencies into a conda environment like here https://tovotu.de/dev/524-Invidious-auf-einem-Uberspace-einrichten/ (unfortunately, only in German).

My setup:
```
$ uname -a
Linux cordelia.uberspace.de 3.10.0-1127.10.1.el7.x86_64 #1 SMP Wed Jun 3 14:28:03 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ crystal version
Crystal 0.35.0 [3c48f311f] (2020-06-09)

LLVM: 8.0.0
Default target: x86_64-unknown-linux-gnu
$ crystal env 
CRYSTAL_CACHE_DIR=/home/user8297/.cache/crystal
CRYSTAL_PATH=lib:/home/user8297/opt/crystal/bin/../share/crystal/src
CRYSTAL_VERSION=0.35.0
CRYSTAL_LIBRARY_PATH=/home/user8297/opt/crystal/bin/../lib/crystal/lib
CRYSTAL_OPTS=''
$ $CC --version
x86_64-conda_cos6-linux-gnu-cc (crosstool-NG 1.23.0.449-a04d0) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by bbugh** [comment]
I am able to reproduce this with 100% consistency while trying to [set up mint](https://github.com/mint-lang/mint/blob/master/CONTRIBUTING.md) on the crystal 1.0.0 docker image.

```
$ docker run -t -i crystallang/crystal:1.0.0 /bin/bash
$ cd ~
$ git clone https://github.com/mint-lang/mint
$ cd mint
$ shards install
$ apt update
$ apt install pkg-config ubuntu-dev-tools
$ mkdir ~/.bin
$ make development
```

It will then die:

```
root@0cabd58b0506:~/mint# make development
shards build --error-on-warnings --error-trace --progress
Dependencies are satisfied
Building: mint
Error target mint failed to compile:
cc: error: _main.o: No such file or directory
cc: error: S-lice40U-I-nt841.o: No such file or directory
cc: error: P-ointer40U-I-nt841.o: No such file or directory
cc: error: A-rgumentE-rror.o: No such file or directory
cc: error: E-xception5858C-allS-tack.o: No such file or directory
cc: error: A-rray40P-ointer40V-oid4141.o: No such file or directory
cc: error: P-ointer40P-ointer40V-oid4141.o: No such file or directory
cc: error: I-nt32.o: No such file or directory
cc: error: F-loat64.o: No such file or directory
```

**Comment by straight-shoota** [comment]
I am unable to reproduce with these exact same steps.

**Comment by bbugh** [comment]
Isn't coding wonderful? 😂 I'm on a week old Mac Big Sur 11.3, if that matters. To debug further:

1. I tried using `ubuntu:20.04" (adf73ca0) image instead of the crystal one, and installed everything manually, and it still had the same problem. 
1. I did a full boot to a Linux Mint live usb (v20.2) and it worked fine
1. I booted back to the Mac and tried running the same steps I posted above with the `linuxmintd/mint20.2-amd64` image instead. It also failed in the same spot.
1. I finally tried `crystallang/crystal:nightly-alpine-build` (0f7866bb39f0) with the equivalent steps and it worked without any issues. ✅

It seems like it might be a debian thing?

**Comment by straight-shoota** [comment]
I would suspect it has something to do with the file system. Or the underlying disc device.

**Comment by mloughran** [comment]
I landed here while debugging a similar issue. Builds (for certain targets – presumably linked to compilation complexity) were failing within a docker container limited to 2GB RAM. The solution was to increase Docker's RAM allowance. If the root cause here is indeed OOM, it would be nice if the error could be more helpful :)

**Comment by straight-shoota** [comment]
Indeed, I have recently encountered this as well and I'm pretty sure it was because there was not enough memory available.

**Comment by atlantis** [comment]
Ooof I'm running into this sporadically trying to build on arm64 IoT device (embedded crystal ftw)... I only have 2GB to work with on the device so cranking up the memory limit isn't an option unfortunately...

**Comment by straight-shoota** [comment]
@atlantis Unfortunately, the Crystal compiler is quite memory hungry, so it doesn't run well in confined environments such as embedded devices.
Your best chance is probably [cross compiling](https://crystal-lang.org/reference/1.6/syntax_and_semantics/cross-compilation.html) from a more potent host system.

**Comment by atlantis** [comment]
Yup! I'm using [Balena](https://www.balena.io/) and the production builds do that, but it's also nice to be able to compile on the device in devmode... crystal works surprisingly well overall on a constrained device.
